### PR TITLE
Fix adding duplicate volumes

### LINKED ISSUE #766: Pods of the InferenceService are not created when using GCS  ServiceAccount credentials
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
I created an `InferenceService` that uses GCS, but a pod is not created.

```
gcloud iam service-accounts keys create gcp-sa-credentials.json \
  --iam-account [SA-NAME]@[PROJECT-ID].iam.gserviceaccount.com
  
  
kubectl -n [NAMESPACE] create secret generic user-gcp-sa \
  --from-file=gcloud-application-credentials.json=gcp-sa-credentials.json
```
ServiceAccount
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: gcs-sa
secrets:
- name: user-gcp-sa
```

inference-service.yaml
```
apiVersion: "serving.kubeflow.org/v1alpha2"
kind: "InferenceService"
metadata:
  name: "mnist-gcs"
spec:
  default:
    predictor:
      serviceAccountName: gcs-sa
      tensorflow:
        storageUri: "gs://my-bucket/mnist/model"
```
Revision -> Deployment -> Replicaset (O)-> Pod(X)

Replicaset error
```
status:
  conditions:
  - lastTransitionTime: "2020-04-05T06:14:27Z"
    message: 'Pod "mnist-gcs-predictor-default-dgsss-deployment-7dd6d67df4-mtfkr"
      is invalid: spec.volumes[4].name: Duplicate value: "user-gcp-sa"'
    reason: FailedCreate
    status: "True"
    type: ReplicaFailure
  observedGeneration: 1
  replicas: 0
```

The `user-gcp-sa` volume is defined in the deployment's pod spec, but it seems to be a problem that occurs when the `user-gcp-sa` volume is duplicated by the pod mutator when the pod is created.

**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version: 
- Knative Version: 
- KFServing Version: 0.2.2
- Kubeflow version: 1.0.1
- Minikube version:
- Kubernetes version: (use `kubectl version`): 1.15.1
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by kangwoo**
**What this PR does / why we need it**:
Fixed duplicate volume being added when creating an `InferenceService` using GCS.

**Which issue(s) this PR fixes** 
Fixes #766 

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by kangwoo** [comment]
@googlebot I signed it!

**Comment by kangwoo** [comment]
@googlebot I signed it!

**Comment by kangwoo** [comment]
Fix conda download failure when building a pytorch server image
Error
```
Step 4/14 : RUN curl -o ~/miniconda.sh -O  https://repo.continuum.io/miniconda/Miniconda3-4.2.12-Linux-x86_64.sh  &&      chmod +x ~/miniconda.sh &&      ~/miniconda.sh -b -p /opt/conda &&      rm ~/miniconda.sh &&      /opt/conda/bin/conda install conda-build &&      /opt/conda/bin/conda create -y --name pytorch-py37 python=3.7.3 numpy pyyaml scipy ipython mkl &&      /opt/conda/bin/conda clean -ya
 ---> Running in 628c1c4a6401
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
/bin/sh: 1: /opt/conda/bin/conda: not found
The command '/bin/sh -c curl -o ~/miniconda.sh -O  https://repo.continuum.io/miniconda/Miniconda3-4.2.12-Linux-x86_64.sh  &&      chmod +x ~/miniconda.sh &&      ~/miniconda.sh -b -p /opt/conda &&      rm ~/miniconda.sh &&      /opt/conda/bin/conda install conda-build &&      /opt/conda/bin/conda create -y --name pytorch-py37 python=3.7.3 numpy pyyaml scipy ipython mkl &&      /opt/conda/bin/conda clean -ya' returned a non-zero code: 127
```

Cause
```
curl -v -O https://repo.continuum.io/miniconda/Miniconda3-4.2.12-Linux-x86_64.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 104.18.200.79...
* Connected to repo.continuum.io (104.18.200.79) port 443 (#0)
* found 148 certificates in /etc/ssl/certs/ca-certificates.crt
* found 592 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_ECDSA_AES_128_GCM_SHA256
* 	 server certificate verification OK
* 	 server certificate status verification SKIPPED
* 	 common name: sni.cloudflaressl.com (matched)
* 	 server certificate expiration date OK
* 	 server certificate activation date OK
* 	 certificate public key: EC
* 	 certificate version: #3
* 	 subject: C=US,ST=CA,L=San Francisco,O=Cloudflare\, Inc.,CN=sni.cloudflaressl.com
* 	 start date: Mon, 06 Apr 2020 00:00:00 GMT
* 	 expire date: Fri, 09 Oct 2020 12:00:00 GMT
* 	 issuer: C=US,ST=CA,L=San Francisco,O=CloudFlare\, Inc.,CN=CloudFlare Inc ECC CA-2
* 	 compression: NULL
* ALPN, server accepted to use http/1.1
> GET /miniconda/Miniconda3-4.2.12-Linux-x86_64.sh HTTP/1.1
> Host: repo.continuum.io
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Date: Fri, 10 Apr 2020 00:18:18 GMT
< Transfer-Encoding: chunked
< Connection: keep-alive
< Cache-Control: max-age=3600
< Expires: Fri, 10 Apr 2020 01:18:18 GMT
< Location: https://repo.anaconda.com/miniconda/Miniconda3-4.2.12-Linux-x86_64.sh
< Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< Server: cloudflare
< CF-RAY: 58184a31cfcde9f4-ICN
< 
{ [5 bytes data]
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
```

**Comment by yuzisun** [comment]
Nice fix @kangwoo !

File: pkg/controller/inferenceservice/resources/credentials/service_account_credentials.go [hunk 1]
```diff
@@ -120,3 +120,12 @@ func (c *CredentialBuilder) CreateSecretVolumeAndEnv(namespace string, serviceAc
...
 }
+
+func appendVolumeIfNotExists(slice []v1.Volume, volume v1.Volume) []v1.Volume {
+	for _, ele := range slice {
+		if ele.Name == volume.Name {
+			return slice
+		}
+	}
+	return append(slice, volume)
+}
```
Comment (yuzisun) [inline pkg/controller/inferenceservice/resources/credentials/service_account_credentials.go:131]: nit: newline

File: pkg/controller/inferenceservice/resources/credentials/service_account_credentials_test.go [hunk 2]
```diff
@@ -432,3 +432,123 @@ func TestAzureCredentialBuilder(t *testing.T) {
...
+	for name, scenario := range scenarios {
+		volumes := appendVolumeIfNotExists(scenario.volumes, scenario.volume)
+
+		if diff := cmp.Diff(scenario.expectedVolumes, volumes); diff != "" {
+			t.Errorf("Test %q unexpected volume (-want +got): %v", name, diff)
+		}
+	}
+
+
+}
```
Comment (yuzisun) [inline pkg/controller/inferenceservice/resources/credentials/service_account_credentials_test.go:554]: nit: newline

File: python/pytorch.Dockerfile [hunk 3]
```diff
@@ -11,7 +11,7 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
          libpng-dev && \
      rm -rf /var/lib/apt/lists/*
 
-RUN curl -o ~/miniconda.sh -O  https://repo.continuum.io/miniconda/Miniconda3-4.2.12-Linux-x86_64.sh  && \
+RUN curl -L -o ~/miniconda.sh -O  https://repo.continuum.io/miniconda/Miniconda3-4.2.12-Linux-x86_64.sh  && \
```
Comment (yuzisun) [inline python/pytorch.Dockerfile:14]: you might need to rebase the fix

File: pkg/controller/inferenceservice/resources/credentials/service_account_credentials.go [hunk 4]
```diff
@@ -120,3 +120,12 @@ func (c *CredentialBuilder) CreateSecretVolumeAndEnv(namespace string, serviceAc
 
 	return nil
 }
+
+func appendVolumeIfNotExists(slice []v1.Volume, volume v1.Volume) []v1.Volume {
```
Comment (yuzisun) [inline pkg/controller/inferenceservice/resources/credentials/service_account_credentials.go:124]: Can we move this to the utils package?

File: python/pytorch.Dockerfile (hunk 3 above)
Comment (kangwoo) [reply python/pytorch.Dockerfile]: Ok. I'll rebase.

File: pkg/controller/inferenceservice/resources/credentials/service_account_credentials.go (hunk 4 above)
Comment (kangwoo) [reply pkg/controller/inferenceservice/resources/credentials/service_account_credentials.go]: I'll move to the utils package

**Pull request merged**